### PR TITLE
Emit warning for complex*4 x

### PR DESCRIPTION
**Pull request by Kishan-Ved**
Fixes: https://github.com/lfortran/lfortran/issues/2104

### CODE REVIEW

**Comment by Kishan-Ved** [comment]
The check seem to fail too soon, what's the reason?

File: src/lfortran/parser/parser.yy [hunk 1]
```diff
@@ -1468,6 +1468,7 @@ var_type
     | KW_COMPLEX { $$ = ATTR_TYPE(Complex, @$); }
     | KW_COMPLEX "(" kind_arg_list ")" { $$ = ATTR_TYPE_KIND(Complex, $3, @$); }
     | KW_COMPLEX "*" TK_INTEGER { $$ = ATTR_TYPE_INT(Complex, DIV2($3), @$); }
+    | KW_COMPLEX "*" TK_INTEGER { $$ = ATTR_TYPE_INT(Complex, $3, @$);  WARN_COMPLEXSTAR($3, @$);}
```
Comment (certik) [inline src/lfortran/parser/parser.yy:1471]: I think you have to do it like this:
```suggestion
    | KW_COMPLEX "*" TK_INTEGER { $$ = ATTR_TYPE_INT(Complex, DIV2($3), @$); WARN_COMPLEXSTAR($3, @$); }
```

**Comment by Kishan-Ved** [comment]
Everything works, thanks! @certik 
Good to be merged?

**Comment by Pranavchiku** [comment]
@HarshitaKalani can you review this?

File: tests/reference/asr-implicit10-23a5156.stderr [hunk 2]
```diff
@@ -3,3 +3,15 @@ style suggestion: Use real(8) instead of real*8
...
+ --> tests/implicit10.f90:8:1
+  |
+8 | complex*8 :: l, m
+  | ^^^^^^^^^ help: write this as 'complex(8)'
+
+style suggestion: Use complex(8) instead of complex*8
+ --> tests/implicit10.f90:9:1
+  |
+9 | complex *16 :: n
+  | ^^^^^^^^^^^ help: write this as 'complex(8)'
```
Comment (HarshitaKalani) [inline tests/reference/asr-implicit10-23a5156.stderr:17]: Here, the style suggestion is wrong. It should be `Use complex(16) instead of complex*16`. Same with help: write this as 'complex(16)'.

File: integration_tests/warning_complexstar.f90 [hunk 3]
```diff
@@ -0,0 +1,3 @@
+program main
```
Comment (HarshitaKalani) [inline integration_tests/warning_complexstar.f90:1]:
```suggestion
program warning_complexstar
```

**Comment by Kishan-Ved** [comment]
I think everything's resolved now and good to go, thanks @HarshitaKalani !

File: src/lfortran/parser/semantics.h [hunk 4]
```diff
@@ -1691,13 +1691,28 @@ void add_ws_warning(const Location &loc,
                 {loc},
                 msg2);
                 
+        } else if (end_token == yytokentype::KW_COMPLEX) {
+                if(a_kind == 16){
```
Comment (certik) [inline src/lfortran/parser/semantics.h:1695]:
```suggestion
                if (a_kind == 16) {
```

File: src/lfortran/parser/semantics.h [hunk 5]
```diff
@@ -1691,13 +1691,28 @@ void add_ws_warning(const Location &loc,
...
+                        diagnostics.parser_style_label(
+                        "Use complex(16) instead of complex*16",
+                        {loc},
+                        "help: write this as 'complex(16)'");
+                } else{
```
Comment (certik) [inline src/lfortran/parser/semantics.h:1700]:
```suggestion
                } else {
```

File: src/lfortran/parser/semantics.h [hunk 6]
```diff
@@ -1691,13 +1691,28 @@ void add_ws_warning(const Location &loc,
...
+                        diagnostics.parser_style_label(
+                        "Use complex(8) instead of complex*8",
+                        {loc},
+                        "help: write this as 'complex(8)'");
```
Comment (certik) [inline src/lfortran/parser/semantics.h:1704]: Use 4 spaces of indentation everywhere.

File: tests/reference/llvm-derived_types_32-4684b97.stderr [hunk 7]
```diff
@@ -0,0 +1 @@
+code generation error: Only integer type is supported for string write(..) for now
```
Comment (certik) [inline tests/reference/llvm-derived_types_32-4684b97.stderr:1]: This seems like some spurious error, unrelated to this PR?

File: integration_tests/CMakeLists.txt [hunk 8]
```diff
@@ -1302,4 +1302,4 @@ RUN(NAME types_20 LABELS gfortran llvm)
 RUN(NAME arrayprint_01 LABELS gfortran llvm)
 RUN(NAME print_arr_05 LABELS gfortran llvm)
 
-RUN(NAME binop_01 LABELS gfortran llvm)
+RUN(NAME binop_01 LABELS gfortran llvm)
```
Comment (certik) [inline integration_tests/CMakeLists.txt:1305]: This change can be removed. If white space changes are needed, let's do them in separate PRs.

**Review by certik — COMMENTED** [review]
Awesome, thanks for the PR! I left some minor comments.

File: tests/reference/llvm-derived_types_32-4684b97.stderr (hunk 7 above)
Comment (Kishan-Ved) [reply tests/reference/llvm-derived_types_32-4684b97.stderr:1]: Yes I don't think this is related to this PR.

File: integration_tests/CMakeLists.txt (hunk 8 above)
Comment (Kishan-Ved) [reply integration_tests/CMakeLists.txt:1305]: I didn't change anything in this file, I don't know why this is showing.

Comment (certik) [reply integration_tests/CMakeLists.txt:1305]: You probably saved it and your editor changes the end of file handling. Just do:
```
git checkout main -- integration_tests/CMakeLists.txt
```

File: tests/reference/llvm-derived_types_32-4684b97.stderr (hunk 7 above)
Comment (certik) [reply tests/reference/llvm-derived_types_32-4684b97.stderr:1]: Just remove the file.

File: integration_tests/CMakeLists.txt (hunk 8 above)
Comment (Kishan-Ved) [reply integration_tests/CMakeLists.txt]: I did this, it deleted 4 lines from the file.

Comment (Kishan-Ved) [reply integration_tests/CMakeLists.txt]: I resolved the conflicts that appeared.

File: integration_tests/CMakeLists.txt [hunk 9]
```diff
@@ -832,9 +832,6 @@ RUN(NAME derived_types_27 LABELS gfortran llvm fortran)
 RUN(NAME derived_types_28 LABELS gfortran llvm)
 RUN(NAME derived_types_29 LABELS gfortran llvm fortran)
 RUN(NAME derived_types_30 LABELS gfortran llvm)
-RUN(NAME derived_types_31 LABELS gfortran llvm)
-RUN(NAME derived_types_32 LABELS gfortran)
-RUN(NAME derived_types_33 LABELS gfortran llvm)
```
Comment (certik) [inline integration_tests/CMakeLists.txt:837]: Return these back in.

**Comment by certik** [comment]
Left a comment above. Otherwise this looks good.

File: src/lfortran/parser/semantics.h [hunk 10]
```diff
@@ -1691,13 +1691,30 @@ void add_ws_warning(const Location &loc,
...
+                            {loc},
+                            "help: write this as 'complex(16)'"
+                        );
+                } else {
+                        diagnostics.parser_style_label(
+                            "Use complex(8) instead of complex*8",
+                            {loc},
+                            "help: write this as 'complex(8)'"
+                        );
+                }
```
Comment (certik) [inline src/lfortran/parser/semantics.h:1707]: Use 4 spaces indentation.

File: src/lfortran/parser/semantics.h [hunk 11]
```diff
@@ -1691,13 +1691,30 @@ void add_ws_warning(const Location &loc,
...
+                                "Use complex(16) instead of complex*16",
+                                {loc},
+                                "help: write this as 'complex(16)'"
+                        );
+                } else {
+                        diagnostics.parser_style_label(
+                                "Use complex(8) instead of complex*8",
+                                {loc},
+                                "help: write this as 'complex(8)'"
+                        );
```
Comment (certik) [inline src/lfortran/parser/semantics.h:1706]: Just use 4 spaces for indentation. I think you are using 8 spaces.

Comment (Kishan-Ved) [reply src/lfortran/parser/semantics.h:1706]: I used 4 spaces in the previous commit, in this commit, I used the spaces format followed in the lines above. Please specify how this has to be done.

Comment (certik) [reply src/lfortran/parser/semantics.h:1706]: Neither commit used 4 spaces. Maybe you used tabs?

Simply ensure exactly 4 spaces are used for indentation, not 8, not tabs.

Comment (Kishan-Ved) [reply src/lfortran/parser/semantics.h]: Fixed in the latest commit. I hope this is correct.

**Comment by Kishan-Ved** [comment]
I was trying to resolve conflicts and I guess it did update a few files, but not all. Someone please help with this.

File: integration_tests/CMakeLists.txt (hunk 9 above)
Comment (certik) [reply integration_tests/CMakeLists.txt:837]: Do:
```
git checkout main -- integration_tests/CMakeLists.txt
git commit
```

File: src/lfortran/parser/semantics.h [hunk 12]
```diff
@@ -1691,13 +1691,28 @@ void add_ws_warning(const Location &loc,
                 {loc},
                 msg2);
```
Comment (certik) [inline src/lfortran/parser/semantics.h:1693]: Remove the empty line:
```suggestion
```

File: src/lfortran/parser/semantics.h [hunk 13]
```diff
@@ -1691,13 +1691,28 @@ void add_ws_warning(const Location &loc,
...
+                diagnostics.parser_style_label(
+                "Use complex(16) instead of complex*16",
+                {loc},
+                "help: write this as 'complex(16)'");
+            } else {
+                diagnostics.parser_style_label(
+                "Use complex(8) instead of complex*8",
+                {loc},
+                "help: write this as 'complex(8)'");
+                }
```
Comment (certik) [inline src/lfortran/parser/semantics.h:1705]: I think the indentation here has to be:
```suggestion
            }
```

**Comment by certik** [comment]
I merged the latest main, that took care of most of the issues. Left a few comments to fix up the rest.

**Comment by Kishan-Ved** [comment]
> I merged the latest main, that took care of most of the issues. Left a few comments to fix up the rest.
Fixed, thanks!

**Comment by Kishan-Ved** [comment]
> I merged the latest main, that took care of most of the issues. Left a few comments to fix up the rest.

@certik Sir how was this fixed? Did you do update branch (without rebase)? As I could see a button with 2 options: Update (without rebase) and update (with rebase), but the update with rebase was shown as an invalid (unallowed) action.

**Review by certik — APPROVED** [review]
This is fine, thanks!

Note that none of the `integration_tests/warning_*.f90` files seem to be tested, since I didn't find any mention of them in any of our tests. This can be addressed in subsequent PRs.

This PR change is tested by the reference tests, so it's good to merge.
